Invariant attribute not transpiled to resulting Metal shader

tadeuzagallo opened the pull request:
#### 13b646f1849dbb58b8348600b12213ddcf69d16a
<pre>
Invariant attribute not transpiled to resulting Metal shader
<a href="https://bugs.webkit.org/show_bug.cgi?id=297397">https://bugs.webkit.org/show_bug.cgi?id=297397</a>

Reviewed by Mike Wyrzykowski.

The issue required three related fixes:
- The code generator previously just ignored invariant attributes, where it should
  emit `[[invariant]]`
- Attributes on return types were ignored, and in order to correctly emit them we
  need to wrap the return value in a struct
- Our existing mechanism for wrapping the return value, which was previously only
  used for fragment outputs, would only wrap the result if the type was a simple
  identifier, ignoring elaborated types such as `vecN&lt;T&gt;` and `array&lt;T&gt;`

* Source/WebGPU/WGSL/AST/ASTArrayTypeExpression.h:
(WGSL::AST::ArrayTypeExpression::maybeElementType const): Deleted.
(WGSL::AST::ArrayTypeExpression::maybeElementCount const): Deleted.
(WGSL::AST::ArrayTypeExpression::ArrayTypeExpression): Deleted.
* Source/WebGPU/WGSL/AST/ASTStructure.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::checkReturnType):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/WGSLShaderModule.h:
(WGSL::ShaderModule::replace):
* Tools/TestWebKitAPI/Tests/WGSL/MetalCompilationTests.mm:
(TestWGSLAPI::TEST_F(WGSLMetalCompilationTests, Attributes)):
* Tools/TestWebKitAPI/Tests/WGSL/MetalGenerationTests.cpp:
(TestWGSLAPI::TEST(WGSLMetalGenerationTests, RedFrag)):

Canonical link: <a href="https://commits.webkit.org/299494@main">https://commits.webkit.org/299494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b0a46a3fd418832da2b961975bcd3705b056605

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119187 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38869 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29520 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125413 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71250 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5e49a08c-5eb6-48b1-a646-2ec68c403d9a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121065 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47450 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90544 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-multicol/crashtests/inline-float-parallel-flow.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c21ca4cc-d2e5-4719-9357-68544f098033) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31544 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106842 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70949 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/18a557b3-b92b-42ad-8432-1b57ec749827) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30588 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24957 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69062 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100990 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128428 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34836 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99102 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46461 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103058 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98882 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25142 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44351 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22354 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42673 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45964 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51644 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45431 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48779 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47121 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (changes requested):
actually it looks like this causes breaking changes in the layout tests, see https://ews-build.s3-us-west-2.amazonaws.com/macOS-Sonoma-Release-WK2-Tests-EWS/a17c3e0c-23474/fast/webgpu/nocrash/fuzz-286564-pretty-diff.html
```
program_source:42:33: error: 'color' attribute parameter is out of bounds: must be between 0 and 7
 88    vec<int, 4> __value [[color(200)]];
```

tadeuzagallo committed:
Invariant attribute not transpiled to resulting Metal shader
https://bugs.webkit.org/show_bug.cgi?id=297397

Reviewed by Mike Wyrzykowski.

The issue required three related fixes:
- The code generator previously just ignored invariant attributes, where it should
  emit `[[invariant]]`
- Attributes on return types were ignored, and in order to correctly emit them we
  need to wrap the return value in a struct
- Our existing mechanism for wrapping the return value, which was previously only
  used for fragment outputs, would only wrap the result if the type was a simple
  identifier, ignoring elaborated types such as `vecN<T>` and `array<T>`

* Source/WebGPU/WGSL/AST/ASTArrayTypeExpression.h:
(WGSL::AST::ArrayTypeExpression::maybeElementType const): Deleted.
(WGSL::AST::ArrayTypeExpression::maybeElementCount const): Deleted.
(WGSL::AST::ArrayTypeExpression::ArrayTypeExpression): Deleted.
* Source/WebGPU/WGSL/AST/ASTStructure.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::checkReturnType):
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/WGSLShaderModule.h:
(WGSL::ShaderModule::replace):
* Tools/TestWebKitAPI/Tests/WGSL/MetalCompilationTests.mm:
(TestWGSLAPI::TEST_F(WGSLMetalCompilationTests, Attributes)):
* Tools/TestWebKitAPI/Tests/WGSL/MetalGenerationTests.cpp:
(TestWGSLAPI::TEST(WGSLMetalGenerationTests, RedFrag)):

Canonical link: https://commits.webkit.org/299494@main
